pkg/controller/validatingadmissionpolicystatus/controller.go

```diff
@@ -0,0 +1,146 @@
+	// Type of params is defined in policy.Spec.ParamsKind
+	// Types of object are calculated from policy.Spec.MatchingConstraints
+	typeChecker *validatingadmissionpolicy.TypeChecker
+}
+
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+
+	if !cache.WaitForNamedCacheSync(ControllerName, ctx.Done(), c.policySynced) {
+		return
+	}
+
+	defer c.policyQueue.ShutDown()
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, c.syncPolicy, 0)
+	}
+
+	<-ctx.Done()
+}
+
+func NewController(policyInformer informerv1alpha1.ValidatingAdmissionPolicyInformer, policyClient admissionregistrationv1alpha1.ValidatingAdmissionPolicyInterface, typeChecker *validatingadmissionpolicy.TypeChecker) (*Controller, error) {
+	c := &Controller{
+		policyInformer: policyInformer,
+		policyQueue:    workqueue.NewRateLimitingQueueWithConfig(workqueue.DefaultControllerRateLimiter(), workqueue.RateLimitingQueueConfig{Name: ControllerName}),
+		policyClient:   policyClient,
+		typeChecker:    typeChecker,
+	}
+	reg, err := policyInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueuePolicy(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.enqueuePolicy(newObj)
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	c.policySynced = reg.HasSynced
+	return c, nil
+}
+
+func (c *Controller) enqueuePolicy(policy any) {
+	if policy, ok := policy.(*v1alpha1.ValidatingAdmissionPolicy); ok {
+		// policy objects are cluster-scoped, no point include its namespace.
+		key := policy.ObjectMeta.Name
+		if key == "" {
+			utilruntime.HandleError(fmt.Errorf("cannot get name of object %v", policy))
+		}
+		c.policyQueue.AddRateLimited(key)
+	}
+}
+
+func (c *Controller) syncPolicy(ctx context.Context) {
+	key, shutdown := c.policyQueue.Get()
+	if shutdown {
+		return
+	}
+	defer c.policyQueue.Done(key)
+	err := func() error {
+		key, ok := key.(string)
+		if !ok {
+			return fmt.Errorf("expect a string but got %v", key)
+		}
+		policy, err := c.policyInformer.Lister().Get(key)
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				// If not found, the policy is being deleting.
+				return c.reconcile(ctx, nil)
+			}
+			return err
+		}
+		return c.reconcile(ctx, policy)
+	}()
+	if err != nil {
+		utilruntime.HandleError(err)
+	}
+	// do not retry, forget anyway
+	c.policyQueue.Forget(key)
+}
+
+func (c *Controller) reconcile(ctx context.Context, policy *v1alpha1.ValidatingAdmissionPolicy) error {
+	if policy == nil {
+		return nil
+	}
+	if policy.Generation <= policy.Status.ObservedGeneration {
+		return nil
+	}
+	warnings := c.typeChecker.Check(policy)
+	status := policy.Status.DeepCopy()
+	status.ObservedGeneration = policy.Generation
+	status.TypeChecking = &v1alpha1.TypeChecking{ExpressionWarnings: warnings}
+	newPolicy := policy.DeepCopy()
+	newPolicy.Status = *status
+	_, err := c.policyClient.UpdateStatus(ctx, newPolicy, metav1.UpdateOptions{})
```

Comment:
Can we use ApplyStatus instead of UpdateStatus?

Reply:
Good catch. This is a good opportunity to use apply.

Reply:
Updated the PR to avoid deepcopying and instead use an apply.